src/lib/mongoBackend/mongoSubCache.cpp

```diff
@@ -0,0 +1,1204 @@
+    cSubP = cSubP->next;
+  }
+
+  LM_T(LmtMongoSubCache, ("--------------------------------"));
+}
+
+
+
+/* ****************************************************************************
+*
+* mongoSubCacheItemRemove - 
+*/
+int mongoSubCacheItemRemove(CachedSubscription* cSubP)
+{
+  CachedSubscription* current = mongoSubCache.head;
+  CachedSubscription* prev    = NULL;
+
+  LM_T(LmtMongoSubCache, ("in mongoSubCacheItemRemove, trying to remove '%s'", cSubP->subscriptionId));
+  while (current != NULL)
+  {
+    if (current == cSubP)
+    {
+      if (cSubP == mongoSubCache.head)  mongoSubCache.head = cSubP->next;
+      if (cSubP == mongoSubCache.tail)  mongoSubCache.tail = prev;
+      if (prev != NULL)                 prev->next         = cSubP->next;
+
+      LM_T(LmtMongoSubCache, ("in mongoSubCacheItemRemove, REMOVING '%s'", cSubP->subscriptionId));
+      ++mongoSubCache.noOfRemoves;
+
+      cachedSubscriptionDestroy(cSubP);
+      delete cSubP;
+      return 0;
+    }
+
+    prev    = current;
+    current = current->next;
+  }
+
+  LM_E(("Runtime Error (item to remove from sub-cache not found)"));
+  return -1;
+}
+
+
+
+/* ****************************************************************************
+*
+* mongoSubCacheRefresh -
+*
+* 1. Empty cache
+* 2. Lookup all subscriptions in the database 
+* 3. Insert them again in the cache (with fresh data from database)
+*/
+static void mongoSubCacheRefresh(std::string database)
+{
+  LM_T(LmtMongoSubCache, ("Refreshing subscription cache for DB '%s'", database.c_str()));
+
+  BSONObj                   query      = BSON("conditions.type" << "ONCHANGE" << CSUB_ENTITIES "." CSUB_ENTITY_ISPATTERN << "true");
+  DBClientBase*             connection = getMongoConnection();
+  auto_ptr<DBClientCursor>  cursor;
+
+  std::string tenant = tenantFromDb(database);
+  try
+  {
+    cursor = connection->query(getSubscribeContextCollectionName(tenant).c_str(), query);
+
+    /*
+     * We have observed that in some cases of DB errors (e.g. the database daemon is down) instead of
+     * raising an exception, the query() method sets the cursor to NULL. In this case, we raise the
+     * exception ourselves
+     */
+    if (cursor.get() == NULL)
+    {
+      throw DBException("Null cursor from mongo (details on this is found in the source code)", 0);
+    }
+    releaseMongoConnection(connection);
+
+    LM_I(("Database Operation Successful (%s)", query.toString().c_str()));
+  }
+  catch (const DBException &e)
+  {
+    releaseMongoConnection(connection);
+    LM_E(("Database Error (DBException: %s)", e.what()));
+    return;
+  }
+  catch (...)
+  {
+    releaseMongoConnection(connection);
+    LM_E(("Database Error (generic exception)"));
+    return;
+  }
+
+  // Call the treat function for each subscription
+  int subNo = 0;
+  while (cursor->more())
+  {
+    BSONObj sub = cursor->next();
+
+    mongoSubCacheItemInsert(tenant.c_str(), sub);
```

Comment:
Right. However, probably not so obvious (note that it has passed unnotices to me... and I supposed to know how the query is done in MongoDB). Probably a comment about "MongoDB query ensures that we only process isPattern subscriptions" just above mongoSubCacheItemInsert() callc ould help.